warehouse/oidc/views.py

```diff
@@ -0,0 +1,86 @@
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import time
+
+from pyramid.view import view_config
+from sqlalchemy import func
+
+from warehouse.macaroons.interfaces import IMacaroonService
+from warehouse.oidc.interfaces import IOIDCProviderService
+from warehouse.packaging.models import Project
+
+
+@view_config(
+    route_name="oidc.mint_token",
+    require_methods=["POST"],
+    renderer="json",
+    require_csrf=False,
+    has_translations=False,
+)
+def mint_token_from_oidc(request):
+    def _invalid(msg):
+        request.response.status = 422
+        return {"error": msg}
+
+    try:
+        body = request.json_body
+    except ValueError:
+        return _invalid("missing body")
+
+    unverified_jwt = body.get("token")
+    if not unverified_jwt:
+        return _invalid("missing token")
+
+    project_name = body.get("project")
+    if not project_name:
+        return _invalid("missing project")
```

Comment:
I'm a little confused why the body can't just be the OIDC token, and why we need a separate field for the project. Can't that just be inferred from the token based on the provider?

Reply:
This is because of our original data model decision: we decided that multiple projects can share an OIDC publisher, since multiple Python packages might live under the same logical source code repository. We probably don't want to mint a token that's scoped for _all_ of them at once, so we need a way to determine which one we should mint for.

Reply:
(I don't see a significant security concern with minting for all of them, since the OIDC JWT bearer is trusted by definition, but that was my reasoning.)

Reply:
I think we probably _do_ want to mint for all of them? I'm assuming that multiple projects that share the same publisher will want to use the same token to publish, instead of making multiple requests varying the project name to get multiple tokens.

Reply:
Makes sense. I'll drop the project from the body and mint for all projects.

Reply:
Does it make sense to retain a body, or can it just be the OIDC token? Presumably everything we need is w/in the token, but leaving the door open to including additional data might help future-proof this.

Reply:
IMO it makes sense to retain the body, even if it adds one or two additional error states (missing, invalid JSON) for the reason you mentioned.

Reply:
Oh, I realize why I did this originally -- I designed the original verification process to look like this:

1. Verify the signature on an untrusted JWT
2. Compare a list of projects against the JWT's claims, seeing if any projects have OIDC providers that match the claims

When really, our verification process should look like this:

1. Verify the signature on an untrusted JWT
2. Check whether we have an `OIDCProvider` corresponding to the now verified (but still untrusted) JWT
3. Return all projects registered against that `OIDCProvider`

That also simplifies our OIDC service interface, which can become something like `find_provider(unverified_jwt: str) -> OIDCProvider`, which in turn can be used to access the project list. I'll refactor towards that.